File: js/openmedkit-flutter/macos/Classes/OpenMedKitFlutterPlugin.swift
import Cocoa
import FlutterMacOS

public final class OpenMedKitFlutterPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "org.openmed.openmedkit_flutter/platform",
            binaryMessenger: registrar.messenger
        )
        let instance = OpenMedKitFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == "prepareModel" else {
            result(FlutterMethodNotImplemented)
            return
        }
        guard
            let arguments = call.arguments as? [String: Any],
            let path = arguments["modelDirectory"] as? String,
            !path.isEmpty,
            !path.contains("://")
        else {
            result(FlutterError(
                code: "invalid_model_directory",
                message: "A local model directory is required.",
                details: nil
            ))
            return
        }
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory),
              isDirectory.boolValue else {
            result(FlutterError(
                code: "model_directory_missing",
                message: "The local model directory is unavailable.",
                details: nil
            ))
            return
        }
        result(URL(fileURLWithPath: path).standardizedFileURL.path)
    }
}

File: js/openmedkit-flutter/android/src/main/kotlin/org/openmed/openmedkit/flutter/OpenMedKitFlutterPlugin.kt
package org.openmed.openmedkit.flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

/** Registers the offline model-path channel and bundled FFI library. */
class OpenMedKitFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        System.loadLibrary("openmed_ffi")
        channel = MethodChannel(
            binding.binaryMessenger,
            "org.openmed.openmedkit_flutter/platform",
        )
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "prepareModel") {
            result.notImplemented()
            return
        }
        val modelDirectory = call.argument<String>("modelDirectory")
        if (modelDirectory.isNullOrBlank() || modelDirectory.contains("://")) {
            result.error("invalid_model_directory", "A local model directory is required.", null)
            return
        }
        val directory = File(modelDirectory)
        if (!directory.isDirectory) {
            result.error("model_directory_missing", "The local model directory is unavailable.", null)
            return
        }
        result.success(directory.absolutePath)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
